Exceptions
----------
Comes under throwable.
Throwable is not a interface but a class for exception handling.
It have two child classes - 1. Exception, 2, Error

                                                                Throwable
                                                        They are Partially Checked
----------------------------------------------------------------------------------------------------------------------------
Exception                                                                                                              Error
---------                                                                                                              -----
Exceptions are caused by code/program.                                                                                 Error are not caused by code/program.
Exceptions are recoverable (handling it).                                                                              Error are non-recoverable (can'r be handled ).
    use try catch.                                                                                                         Caused by lack of system resources. ie. not enough stack memory, network issue, VM error
Exceptions are partially checked, bcz its child classes few are checked few are unchecked.                             All errors are unchecked exceptions

RuntimeException                    IOException                 InterruptedException        RemoteException            VirtualMachineError        LinkageError    AssertionError      ExceptionInInitializerError
----------------                    -----------                 --------------------        ---------------            -------------------        ------------    --------------      ---------------------------
'Unchecked exception'               'Fully Checked exception'   'Fully Checked exception'   'Fully Checked exception'  StackOverFlowError         VerifyError
ArithmaticException                 EndOfFileException                                                                 OutOfMemoryError
NullPointerException                FileNotFoundException                                                              
ClassCachedException
IndexOutOfBoundException
    ArrayIndexOutofBoundException
    StringIndexOutOfBoundException
IllegalArgumentException
    NumberFormatException



Checked vs Unchecked Exception
------------------------------
Checked
-------
Checked exception have 'throws' keyword.
The exceptions which are checked by the compiler for smooth executiuon of the program at runtime. Whether the program is handling or not.
Except runtime exceptions all other exceptions are checked exception.

FileNotFoundException
---------------------
                        import java.io.*;
                        class Test {
                            public static void main(String[] args) {
                                PrintWriter pw = new PrintWriter("abc.txt");
                                pw.println("Hello");
                            }
                        }
error: unreported exception FileNotFoundException must be caught or declare to be thrown.
        at PrintWriter pw = new PrintWriter("abc.txt");
Use try catch or declare throws expetion in class.

IntruptedException
------------------
                    class Test {
                        public static void main(String[] args) {
                                System.out.println("I want to sleep peacefully");
                            Thread.sleep(1000); //every java program compulsary contain 1 thread. i.e. main thread.
                        }
                    }
If main thread entered into the spleep. Java give error. Because other threads will intrupt this.
error: unreported exception IntruptedException must be caught or declare to be thrown.
        at Thread.sleep(1000);


Unchecked
---------
Exceptions which are not checked by compiler.
Rearly occured exceptions.
All errors are unchecked exception, and runtime exceptions is also unchecked exception.

ArithmeticException
-------------------
                    class Test {
                        public static void main(String[] args) {
                            System.out.println(10/0);
                        }
                    }
error: Exception in thread 'main' java.lang.AirthmeticException: /by zero
        at Test.main(Test.java.5)


Partially Checked and Fully Checked Exception
---------------------------------------------
Throwables are Partially Checked Exception.
Exceptions are partially checked, because they are checked but some of its child are unchecked.
Partial Checked are those, whose child exceptions are unchecked.
Fully Checked are those whose parent class is checked and their child exceptions are also checked.


Try-Catch Control Flow
-----------------------
                        statement-0;
                        try {
                            statement-1;
                            statement-2;
                            statement-3;
                        }
                        catch(exception e) {
                            statement-4
                        }
                        statement-5;
case-1 If there is no exception. All the statements will excecute normally except catch block.
            Because no exception raised no need for catch.
case-2 If as exception raised at statement-2 and corresponding catch block matched, statement 4 is executed.
            Because after a exception catch block is executed and after that try block don't execute again, the remaining code will execute after the catch.
case-3 If an exception raised at statement-2 and corresponding catch block not matched. 
            Program will stop abnormally as no handler is present.
case-4 If an exception raised at catch block or after it.
            Beacuse its not part of a try catch block, it is as abnormal termination, program will stop.             

Methods to print exceptions information
---------------------------------------
                        class Test {
                            public static void main(String[] args) {
                                try {
                                    System.out.println(10/0);
                                }
                                catch(ArithmaticException e) {
                                    // e.printStackTrace();
                                    // System.out.println(e);
                                    // System.out.println(e.toString);
                                    System.out.println(e.getMessage());
                                }
                            }
                        }
1. e.printStackTrace()
----------------------
This will tell the name of exception folloewed by it's discription and the stak trace of the error.
By default this is used. By the dfault exception handler.
It has print statement internally so no need for system.out.println.

2. e.toString() or e
--------------------
This will give only name & discription for the error.
These are toString methods so need to print by system.out.println.

3. e.getMessage()
-----------------
This will only give the discription of error.
These are toString methods so need to print by system.out.println.


'Throw' keyword
---------------
Used for Customised, user defined exceptions. Not for pre-defined exception.
To handle our created exception object to the JVM manually.
These are toString methods so need to print by system.out.println.
